tests/convergence/rates.cpp

```diff
@@ -0,0 +1,338 @@
+//               prob 1: H(curl) projection
+//                       (curl u, curl v) + (u,v) = (curl u_exact, curl v) + (u_exact, v)
+//               prob 2: H(div) projection
+//                       (div  u, div  v) + (u,v) = (div  u_exact, div  v) + (u_exact, v)
+//               prob 3: DG discretization for the Poisson problem 
+//                       -Delta u = f
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+using namespace std;
+using namespace mfem;
+
+
+// Exact solution parameters: 
+double sol_s[3] = { -0.32, 0.15, 0.24 };
+double sol_k[3] = { 1.21, 1.45, 1.37 };
+
+// H1
+double u_exact(const Vector &x);
+double rhs_func(const Vector &x);
+void gradu_exact(const Vector &x, Vector &gradu);
+// Vector FE
+void U_exact(const Vector &x, Vector & U);
+// H(curl)
+void curlU_exact(const Vector &x, Vector &curlU);
+// H(div)
+double divU_exact(const Vector &x);
+
+int dim;
+int prob=0;
+
+int main(int argc, char *argv[])
+{
+   // 1. Parse command-line options.
+   const char *mesh_file = "../../data/inline-quad.mesh";
+   int order = 1;
+   bool visualization = 1;
+   int sr = 1;
+   double sigma = -1.0;
+   double kappa = -1.0;
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Mesh file to use.");
+   args.AddOption(&order, "-o", "--order",
+                  "Finite element order (polynomial degree)");
+   args.AddOption(&prob, "-prob", "--problem",
+                  "Problem kind: 0: H1, 1: H(curl), 2: H(div), 3: DG ");
+   args.AddOption(&sigma, "-s", "--sigma",
+                  "One of the two DG penalty parameters, typically +1/-1."
+                  " See the documentation of class DGDiffusionIntegrator.");
+   args.AddOption(&kappa, "-k", "--kappa",
+                  "One of the two DG penalty parameters, should be positive."
+                  " Negative values are replaced with (order+1)^2.");               
+   args.AddOption(&sr, "-sr", "--serial_ref",
+                  "Number of serial refinements.");
+   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+   args.Parse();
+   if (!args.Good())
+   {
+      args.PrintUsage(cout);
+      return 1;
+   }
+   if (prob >3 || prob <0) prob = 0; //default problem = H1
+   if (prob == 3)
+   {
+      if (kappa < 0)
+      {
+         kappa = (order+1)*(order+1);
+      }
+   }
+   args.PrintOptions(cout);
+
+   // 2. Read the (serial) mesh from the given mesh file. 
+   Mesh *mesh = new Mesh(mesh_file, 1, 1);
+   dim = mesh->Dimension();  
+
+   // 4. Refine the serial mesh on all processors to increase the resolution.
+   mesh->UniformRefinement();
+
+   // 7. Define a finite element space on the parallel mesh.
+   FiniteElementCollection *fec=nullptr;
+   switch (prob)
+   {
+      case 0: fec = new H1_FECollection(order,dim);   break;
```

Comment:
Adding an `enum class` or something in this file for the various function types would increase the readability of this code.

Reply:
You think we need more this description ?

https://github.com/mfem/mfem/blob/ab0955357c10a9ef3d60bb8555a810a7212e9616/tests/convergence/rates.cpp#L26-L37

Reply:
The issue is that this description is at the top of the file. This requires a new user to keep this in their head as they are reading through the file. As with the previous issues, this isn't a showstopper and I defer to @tzanio and @v-dobrev 's opinions. I realize they my style preferences are different from the larger MFEM team! 😄

Reply:
We have things like this in several other example already, so I think it is OK to stay as it is.